### ISSUE TITLE
`rpminspect -k` downloads all RPMs on each invocation

### ISSUE THREAD

**Issue by kdudka** [rpminspect/rpminspect#759]
I tried to run `rpminspect` locally.  On each invocation I had to wait minutes until all RPMs were downloaded.  This is not an acceptable overhead while debugging rpminspect false positives to suppress them with `rpminspect.yaml`.  Using options `-k` and `-w` did not help.  A fresh directory was used for the downloaded RPMs each time.  Is this expected?
```
$ rpminspect-redhat --version
rpminspect version 1.10

$ rpminspect-redhat -k -w /tmp/dir -T rpmdeps -v curl-7.76.1-18.el9
Downloading RPM build curl-7.76.1-18.el9
=> curl-7.76.1-18.el9.src.rpm         [######################################]
=> libcurl-minimal-...el9.aarch64.rpm  [######################################]
=> libcurl-7.76.1-18.el9.aarch64.rpm  [######################################]
=> curl-debuginfo-7...el9.aarch64.rpm  [######################################]
=> curl-7.76.1-18.el9.aarch64.rpm     [######################################]
=> curl-minimal-7.7...el9.aarch64.rpm  [######################################]
=> libcurl-minimal-...el9.aarch64.rpm  [######################################]
=> curl-minimal-deb...el9.aarch64.rpm  [######################################]
=> libcurl-debuginf...el9.aarch64.rpm  [######################################]
=> curl-debugsource...el9.aarch64.rpm  [######################################]
=> libcurl-devel-7....el9.aarch64.rpm  [######################################]
=> libcurl-debuginf...el9.ppc64le.rpm  [######################################]
=> curl-minimal-deb...el9.ppc64le.rpm  [######################################]
=> curl-debuginfo-7...el9.ppc64le.rpm  [######################################]
=> libcurl-minimal-...el9.ppc64le.rpm  [######################################]
=> libcurl-minimal-...el9.ppc64le.rpm  [######################################]
=> libcurl-7.76.1-18.el9.ppc64le.rpm  [######################################]
=> curl-7.76.1-18.el9.ppc64le.rpm     [######################################]
=> curl-debugsource...el9.ppc64le.rpm  [######################################]
=> curl-minimal-7.7...el9.ppc64le.rpm  [######################################]
=> libcurl-devel-7....el9.ppc64le.rpm  [######################################]
=> libcurl-debuginf...18.el9.i686.rpm  [######################################]
=> libcurl-minimal-...18.el9.i686.rpm  [######################################]
=> curl-debugsource...18.el9.i686.rpm  [######################################]
=> libcurl-devel-7....18.el9.i686.rpm  [######################################]
=> libcurl-7.76.1-18.el9.i686.rpm     [######################################]
=> curl-7.76.1-18.el9.i686.rpm        [######################################]
=> curl-debuginfo-7...18.el9.i686.rpm  [######################################]
=> libcurl-minimal-...18.el9.i686.rpm  [######################################]
=> curl-minimal-deb...18.el9.i686.rpm  [######################################]
=> curl-minimal-7.7...18.el9.i686.rpm  [######################################]
=> libcurl-minimal-....el9.x86_64.rpm  [######################################]
=> curl-debugsource....el9.x86_64.rpm  [######################################]
=> libcurl-devel-7.....el9.x86_64.rpm  [######################################]
=> libcurl-debuginf....el9.x86_64.rpm  [######################################]
=> curl-debuginfo-7....el9.x86_64.rpm  [######################################]
=> curl-7.76.1-18.el9.x86_64.rpm      [######################################]
=> libcurl-7.76.1-18.el9.x86_64.rpm   [######################################]
=> libcurl-minimal-....el9.x86_64.rpm  [######################################]
=> curl-minimal-deb....el9.x86_64.rpm  [######################################]
=> curl-minimal-7.7....el9.x86_64.rpm  [######################################]
=> libcurl-debuginf...8.el9.s390x.rpm  [######################################]
=> libcurl-minimal-...8.el9.s390x.rpm  [######################################]
=> curl-debugsource...8.el9.s390x.rpm  [######################################]
=> libcurl-devel-7....8.el9.s390x.rpm  [######################################]
=> curl-7.76.1-18.el9.s390x.rpm       [######################################]
=> libcurl-7.76.1-18.el9.s390x.rpm    [######################################]
=> curl-debuginfo-7...8.el9.s390x.rpm  [######################################]
=> libcurl-minimal-...8.el9.s390x.rpm  [######################################]
=> curl-minimal-deb...8.el9.s390x.rpm  [######################################]
=> curl-minimal-7.7...8.el9.s390x.rpm  [######################################]
Running rpmdeps inspection...        FAIL
diagnostics:
[...]
Keeping working directory: /tmp/dir/curl-7.76.1.v7SK2t
```

**Comment by dcantrell** [comment]
The -k option doesn't work that way.  The purpose of -k is more for development to keep the working directory around after rpminspect runs.

If you want to prevent rpminspect from downloading the build each time, you need to first fetch them and place them in your current working directory.  The -f option can do this:

```
rpminspect-redhat -f -v -w . curl-7.76.1-18.el9
```

Then when you run ```rpminspect-redhat -k -w /tmp/dir -T rpmdeps -v curl-7.76.1-18.el9``` from the same directory, rpminspect will get the build from the curl-7.76.1-18.el9 directory containing the build you downloaded.

**Comment by kdudka** [comment]
This worked.  Thanks for the help!  I find the behavior neither intuitive nor documented though.

**Comment by dcantrell** [comment]
It's documented here:
https://rpminspect.readthedocs.io/en/latest/usage.html#workflow-examples

And here:
https://one.redhat.com/rhel-development-guide/#proc_running-rpminspect-locally_assembly_automated-testing

If it needs further clarification, let me know.  I added the -k option initially for debugging.  That's why it's described as keeping the working files.  The -f option was added later at the request of users who wanted to use rpminspect to just fetch builds and feed them in to rpminspect many times over.  I did not recycle the -k option for that functionality.

**Comment by kdudka** [comment]
I find the command-line interface of rpminspect non-standard in multiple ways:
1. Why should users themselves care whether they run rpminspect for the first time for a particular build ID?  I would prefer an idempotent interface that automatically fetches resources as needed and caches them for subsequent runs.
2. When I manually specify a working directory with `-w`, I would expect rpminspect to put everything into the working directory that was specified.  As I understand it, rpminspect creates a new subdirectory with unpredictable name on each invocation.  This does not seem to be documented in any of the resources you refer to.

**Comment by kdudka** [comment]
> The -w option overrides the directory where rpminspect will create all of its working files. That directory will not be used as-is because you can run more than one rpminspect process at a time, so you get a uniquely named temporary directory in the working directory. But this is all internal to rpminspect's functioning.

That is what surprised me because tools like `rpmbuild`, `mock`, etc. do not work this way.

> If you use -f to download a Koji build and specify a working directory with -w, it will download the build directly to that working directory.

One can easily download a build with `brew download-build $NVR`.  There is no need to use `rpminspect` for that.

>  This was requested about 2 years ago when people wanted to first run rpminspect to download a build somewhere and then invoke rpminspect later and specify that locally downloaded build.

I understand how it works but the interface is unfortunate from user's point of view.  One does not need to tell `mock` whether a package is built for the first time or not.  All the caches (root, yum, ccache) work fully transparently.

**Comment by kdudka** [comment]
> Neither multiple rpmbuild processes for the same package nor multiple mock builds for the same package and root can run concurrently.

Both of them can run in parallel if you specify separate working directories for each instance:
```sh
rpmbuild --rebuild curl-7.83.1-1.fc37.src.rpm --define '_topdir /tmp/dir1' &
rpmbuild --rebuild curl-7.83.1-1.fc37.src.rpm --define '_topdir /tmp/dir2' &
mock -r fedora-rawhide-x86_64 --rebuild curl-7.83.1-1.fc37.src.rpm --config-opts=root=/tmp/dir3 &
mock -r fedora-rawhide-x86_64 --rebuild curl-7.83.1-1.fc37.src.rpm --config-opts=root=/tmp/dir4 &
```

> Are you asking for a feature here or just complaining?

Originally I tried to report a bug.  After your first reply I understood how it works.  I still do not understand why it works like this but I am able to script what I need around the current interface.  If you think that rpminspect's interface (and its documentation) is good enough as it is, no further action is needed on this issue.
